### PR TITLE
Cleanup length and bit_length kernels

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-string/src/length.rs [hunk 1]
```diff
@@ -19,135 +19,29 @@
...
-            )
-        };
-        make_array(data)
-    }};
+fn length_impl<P: ArrowPrimitiveType>(
+    offsets: &OffsetBuffer<P::Native>,
+    nulls: Option<&NullBuffer>,
+) -> ArrayRef {
+    let v: Vec<_> = offsets
+        .windows(2)
```
Comment (alamb) [inline arrow-string/src/length.rs:31]: this is very clever 👍

File: arrow-string/src/length.rs [hunk 2]
```diff
@@ -194,28 +98,42 @@ pub fn length(array: &dyn Array) -> Result<ArrayRef, ArrowError> {
...
+        }
+        DataType::Binary => {
+            let list = array.as_binary::<i32>();
+            Ok(bit_length_impl::<Int32Type>(list.offsets(), list.nulls()))
+        }
+        DataType::LargeBinary => {
+            let list = array.as_binary::<i64>();
+            Ok(bit_length_impl::<Int64Type>(list.offsets(), list.nulls()))
+        }
+        DataType::FixedSizeBinary(len) => Ok(Arc::new(Int32Array::new(
```
Comment (alamb) [inline arrow-string/src/length.rs:131]: the support for FixedSizeBinary seems to be new here but there is no test update -- maybe we can add a test for it?

File: arrow-string/src/length.rs [hunk 3]
```diff
@@ -194,28 +98,42 @@ pub fn length(array: &dyn Array) -> Result<ArrayRef, ArrowError> {
...
+            vec![*len * 8; array.len()].into(),
+            array.nulls().cloned(),
+        ))),
         other => Err(ArrowError::ComputeError(format!(
-            "bit_length not supported for {other:?}"
+            "length not supported for {other:?}"
```
Comment (alamb) [inline arrow-string/src/length.rs:136]: I think the prior message was correct as this is the `bit_length` implementation 🤔

**Review by alamb — APPROVED** [review]
A few minor nits, but otherwise this looks like a really nice cleanup @tustvold 

Less unsafe and perhaps faster 👍

File: arrow-string/src/length.rs (hunk 3 above)
Comment (Dandandan) [inline arrow-string/src/length.rs:136]:
```suggestion
            "bit_length not supported for {other:?}"
```

Comment (tustvold) [reply arrow-string/src/length.rs:136]: Yeah this was a copypasta

**Pull request merged**